Setsquare

rolandlo opened the pull request:
This PR (first commit) adds a basic setsquare, which can be moved around, rotated and resized using touch only. It's still in a very early stage and I will work on improving the code quality and extending the feature set. However I think it could already be useful for things like:
- measuring angles and distances (in cm)
- drawing line segments of given length and angles of given degrees
- drawing perpendicular lines

That's why I didn't want to wait longer to submit this PR. So feel free to try it out. Feel free to make any suggestions.

The icon in the toolbar to activate/readjust or hide it looks like a triangle with a semi-circle in it. Press it and use one or two fingers to manipulate the setsquare.

rolandlo commented:
Here is a small illustration: 
![setsquareDemo](https://user-images.githubusercontent.com/40485433/116467264-c8e1a780-a86f-11eb-8c76-936d25e1904f.gif)

LittleHuba reviewed (changes requested):
Interesting PR. I highlighted a few things in the code. In general, I'm not really happy with the input handling. Most of the complex logic for moving/rotation/zooming is in the handler but should be in the view itself. This would cut down on the complexity of the handler. Ideally, the view provides a single method `handle()`, that takes the input events and keeps track of the gestures internally.

rolandlo commented:
> Interesting PR. I highlighted a few things in the code. In general, I'm not really happy with the input handling. Most of the complex logic for moving/rotation/zooming is in the handler but should be in the view itself. This would cut down on the complexity of the handler. Ideally, the view provides a single method `handle()`, that takes the input events and keeps track of the gestures internally.

So would it make sense to create a `SetsquareHandler` class derived from the `AbstractInputHandler` class like the various InputHandler classes for pen/mouse/keyboard/touch, whose `handle()` method could be called from `InputContext::handle()`?

rolandlo commented:
> A nice feature would be to snap the ruler of the setsquare to ends of strokes when they are within a specific offset of it. Also rotating the setsquare around an end of a stroke that the setsquare is snapped to would help with quick-aligning for angles.

I agree. Most likely I will try to keep the functionality in this PR limited though and make it work as solidly as it gets. I can implement more refined behavior like snapping to strokes (I guess only straight ones, i.e. with two points) in a follow-up PR, when the setsquare is tested on various devices.

rolandlo commented:
@LittleHuba Can you please take a look at the latest commit (8425d3720f7a551b74331c3f4961a4584bd07e78) to check if I made proper use of smart pointers? It seems to work, but I have very little experience with smart pointers, so I feel unsure about it.

rolandlo commented:
/azp run create-installers

LittleHuba commented:
I put the comments directly into the commit. You'll find them here 8425d3720f7a551b74331c3f4961a4584bd07e78. Your usage of smart pointers seems right, except for that one case with the if statement. Smart pointers inherently support boolean evaluation. 

You might have already seen the core guidelines for smart pointers. At least you already adhered to them as far as I can see. If you have not yet read them, I recommend that you check them out. https://www.modernescpp.com/index.php/c-core-guidelines-rules-to-smart-pointers

LittleHuba commented:
> Pen input near the smaller two sides of the setsquare could be used for something, but I am not sure what to use it best for, so I decided to just let it draw normally.

How about creating lines from the origin of the setsquare? Like you point your pen at 20° and it proposes a line from the origin to 20°?

rolandlo commented:
> > Pen input near the smaller two sides of the setsquare could be used for something, but I am not sure what to use it best for, so I decided to just let it draw normally.
> 
> How about creating lines from the origin of the setsquare? Like you point your pen at 20° and it proposes a line from the origin to 20°?

Yes, probably that's the best use of it, better than drawing at the short sides of the setsquare like on the long side.

rolandlo commented:
> > > Pen input near the smaller two sides of the setsquare could be used for something, but I am not sure what to use it best for, so I decided to just let it draw normally.
> > 
> > 
> > How about creating lines from the origin of the setsquare? Like you point your pen at 20° and it proposes a line from the origin to 20°?
> 
> Yes, probably that's the best use of it, better than drawing at the short sides of the setsquare like on the long side.

I implemented your idea. Now if you press the mouse or stylus button near one of the short sides of the setsquare, a temporary line is drawn from that point to the setsquare origin. It is added to the layer once the button is released. Between button press and button release the first end point of the line can be modified. If you go inside the semi-circle the angle is not changed any more, only the distance from the origin (so you can draw small line segments with a precise angle).

LittleHuba commented:
Perfect! That is exactly as I imagined this.

rolandlo commented:
I'm experiences some nasty (seemingly random) freezes on X11, which sometimes make all apps irresponsible. It only happens when I annotate a PDF, have the setsquare active and press the buttons on my stylus. On Wayland I don't get those freezes (unless I set `GDK_BACKEND=x11). I haven't figured out what is wrong. The app seems to get hung up in the Gtk main loop.

lxdb commented:
Hi and thanks for the amazing tool!

When merging this PR my surface pen no longer draws anything, but when using the mouse I can normally draw.

Is there any detailed information I could provide to fix this issue?

rolandlo commented:
Hi @lxdb 

Thanks for testing the PR. Haven't got much feedback yet and feedback is very welcome. On my Linux convertible (at least on Wayland) I don't experience any issues. The pen just works fine and the setsquare as well. Are you on Windows 10? Can you take a screenshot of your `input system` panel in menu `Edit > Preferences` and post your `settings.xml` file (that you find in the config folder, see [file locations](https://xournalpp.github.io/guide/file-locations/)). Does the problem also occur, if you do not use the setsquare? On X11 I have some issues, but only when using the setsquare and the stylus buttons and annotating a PDF. Probably you are experiencing a different problem.

lxdb commented:
I always forget the important things.... Yes, I am on WIndows.
![Here](https://user-images.githubusercontent.com/47871822/149576858-b4e9faa9-f006-4843-917a-d5efad6d235b.png) you can see the screenshot of the input panel.
The issue occurs when using xournalpp with setsquare merged, independent of the state of the setsquare.
You can find my settings at https://pastebin.com/sXgAUKAJ . GitHub seems not supporting xml...

rolandlo commented:
/azp run create-installers

LittleHuba commented:
The requested installers were generated successfully. The installers are available at:

- [Windows](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTc1OTEvYXJ0aWZhY3ROYW1lL3dpbmRvd3M1/content?format=zip)
- [macOS](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTc1OTEvYXJ0aWZhY3ROYW1lL21hY09T0/content?format=zip)
- [Debian](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTc1OTEvYXJ0aWZhY3ROYW1lL2xpbnV4LWRlYmlhbi1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 18.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTc1OTEvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0xOC4wNC1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 20.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTc1OTEvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0yMC4wNC1wYWNrYWdlcw2/content?format=zip)

Please be aware that these links are only valid for a limited time (~30 days).

rolandlo commented:
@lxdb Do you get any console messages when running the setsquare version of `xournalpp.exe` (via a MSYS2 or MinGW64 console)? If so, please post them here. 

If there is a bug in my PR causing the issue, then it is most likely either in `SetsquareInputHandler::handleImpl` or in the `InputContext.cpp` changes.

rolandlo commented:
/azp run create-installers

LittleHuba commented:
The requested installers were generated successfully. The installers are available at:

- [Windows](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTgxODcvYXJ0aWZhY3ROYW1lL3dpbmRvd3M1/content?format=zip)
- [macOS](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTgxODcvYXJ0aWZhY3ROYW1lL21hY09T0/content?format=zip)
- [Debian](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTgxODcvYXJ0aWZhY3ROYW1lL2xpbnV4LWRlYmlhbi1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 18.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTgxODcvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0xOC4wNC1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 20.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTgxODcvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0yMC4wNC1wYWNrYWdlcw2/content?format=zip)

Please be aware that these links are only valid for a limited time (~30 days).

rolandlo commented:
The nasty freeze (that only appears under X11, when a PDF is open and the setsquare is displayed) disappears when no text (numbers) is rendered in `Setsquare.cpp`. In particular rotated text seems to make trouble. I have already improved the performance a bit (via clipping), so the issue does not appear every time anymore with the reproducer. An attempt at solving this issue would be to cache the text that needs to be rendered on the Setsquare as much as possible. 

@xournalpp/core Does anyone of you happen to have a better idea about it?

rolandlo commented:
Hurray! I managed to fix the nasty X11-specific bug by replacing `cairo_show_text` by `cairo_text_path`. The latter does not cache glyphs, which would be a disadvantage for lots of text. The text displayed on the setsquare is really short (just some numbers, few digits), so the caching is not particularly useful. I don't get freezes anymore with this replacement. 

Now that the nasty bug is solved, this PR is ready for review. I rewrote history to make it easy for reviewing.
@LittleHuba Would you have time to make the review? You have already looked into the code, so it would be easiest for you.

LittleHuba reviewed (commented):
Thanks for this amazing work! This is a really clean PR that was nice to review!
I have some comments ranging from small notes to some bigger modifications of the overall architecture.

For me, it would be okay if you do this in a follow-up PR (except for the ownership issue). You can also do it in this PR. It's up to you.

rolandlo commented on src/core/model/Setsquare.h:
```diff
@@ -0,0 +1,162 @@
+
+    /**
+     * @brief A setsquare specified by its size, rotation angle and translations in x- and y-directions
+     * @param size half the length of the longest side of the setsquare
```
Maybe it's simplest to rename `size` to `height`. Using the length of its longest side instead will require dividing by 2 at a lot of places in the code.

LittleHuba replied:
You can do that or find a better name for the parameter. Just the way it currently is, makes it very undescriptive.

rolandlo commented on src/core/model/Setsquare.cpp:
```diff
@@ -0,0 +1,318 @@
+    if (std ::abs(radius - std::round(radius)) < .2) {
+        offset = std::max(offset, static_cast<int>(24.0 / radius));
+    }
+    numHmarks = static_cast<int>(std::round((size - 1) * 10.0));
```
Yeah, there were lots of magic numbers. Now it's greatly reduced.

rolandlo commented on src/core/model/Setsquare.h:
```diff
@@ -0,0 +1,162 @@
+    void drawOutline(cairo_t* cr) const;
+    void drawRotation(cairo_t* cr) const;
+    void clipVerticalStripes(cairo_t* cr) const;
+    void clipHorizontalStripes(cairo_t* cr) const;
```
I have added a `Setsquare_Readme.md` in `development/documentation` with pictures explaining everything. The readme also explains how the setsquare handles input.

rolandlo commented:
@LittleHuba Thanks you very much for helpful and speedy review. It's always great to learn from more experienced devs. :smile: 
I have implemented all your feedback except of the refactoring regarding the MVC pattern. I'm not so familiar with programming patterns and will need to read up on that, before I can do the refactoring. This will be part of a new PR. One thing that complicates matters is that there is also a `SetsquareInputHandler` destinct from the `SetsquareController`. I'm not yet sure how they should be related.

I have also tweaked the drawing of the setsquare a bit while making the code more readable. The documentation (Setsquare_Readme.md) should also help anyone who wants to tweak the Setsquare in future. The icons have been rotated 180°, since they should depict the way the setsquare looks when initially activated on a page.

I have already rebased on master. From my side everything seems done. Can you take another look at it (last 3 commits)?

Febbe commented:
To throw in my 50ct, the MVC/MVP pattern is quite outdated. Also I don't like it. There are more modern event driven approaches which are more flexible like flux. But as we use MVC quite often we should stay with it or at least MVP which is sort of the successor of MVC (since we have several controllers we actually use MVP)

LittleHuba commented:
I'll give it a look in a few minutes.

Having a input handler is like splitting the View class of the MVC pattern into two different classes. That should help you understand how this works in your case. Doing that in a follow-up is perfectly fine for me.

@Febbe I agree and would also like to use a different pattern in the long run. But it is even worse to mix patterns. So we should stick to the one we have and think about a switch at all relevant places in an ordered fashion, instead of making the situation more complex. We can open a ticket if you want. But it would make sense to first migrate to GTK4 before starting on this.

LittleHuba reviewed (commented):
Please squash the changes from 538bfba and 5235a29 into the relevant commits where the initial code was added. When that is done, this is good to go. Thanks for your work!

rolandlo commented:
> Please squash the changes from [538bfba](https://github.com/xournalpp/xournalpp/commit/538bfba258b0feb420e2dff8e18e9acd3b553896) and [5235a29](https://github.com/xournalpp/xournalpp/commit/5235a29f7de1623d26896325ac960be12980d127) into the relevant commits where the initial code was added. When that is done, this is good to go. Thanks for your work!

Done! Merging in 48 hours if no objections are raised.
Thanks again @LittleHuba for the speedy review!

lxdb commented:
After testing the merged version, drawing with the Surface Pen on Windows 11 still does not work at all.

Is there any information I could provide for you to be able to fix this?

Otherwise, Xournal++ seems to be unusable since now on Surface Device with Windows.

rolandlo commented:
@lxdb That's pretty unfortunate. I don't understand how that can be. Please provide the console messages, as I already asked for in https://github.com/xournalpp/xournalpp/pull/3082#issuecomment-1013751943

lxdb commented:
@rolandlo Sorry, I didn't see that you asked for it. When running normally, the app does not print any console messages. Is there any debugging option or something else I can turn on to get something?

rolandlo commented:
@lxdb You need a [MSYS2](https://www.msys2.org/) console to get debug messages on Windows. So please install MSYS2 and run xournalpp from the MSYS2 console.

You could also go through the commits and check which of the commits causes the problem for you. Remember to delete the build folder before building again. I would guess that 2fae7378a2aca06dc92a23caacbd23f27c148a43 is the guilty commit and the commit before 9f574681c5d2e8242af9e9013fb02a30d581aad7 is still OK.

lxdb commented:
@rolandlo I used MSYS2 embedded into the Windows Terminal before, so I didn't see the console messages. But when using the MSYS2 console I also don't get anything that would look different from the messages I get when using the 1.1.0-dev build.
I'll try going through the commits.

lxdb commented:
@rolandlo you're right, [daec37a](https://github.com/xournalpp/xournalpp/commit/daec37a83d3a2f498c218f56fb47d1c15cb70f5e) is the last commit working.

rolandlo commented:
> @rolandlo you're right, [daec37a](https://github.com/xournalpp/xournalpp/commit/daec37a83d3a2f498c218f56fb47d1c15cb70f5e) is the last commit working.

Since the problematic commit is really small, it should be easy to debug. Just put some `g_message`s to find out what goes wrong. Specifically checkout the guilty commit and:
- put a `g_message("Event handled by setsquare")` on line 107 of `InputContext.cpp` (right before `return true`) to see, if the event is handled by the `SetsquareInputHandler` (which should not happen if the setsquare is not active). 
- put a `g_message("Device blocked")` on current line 173 and a `g_message("Device unblocked")` on current line 189 of `InputContext.cpp` (compare https://github.com/xournalpp/xournalpp/pull/3082/commits/2fae7378a2aca06dc92a23caacbd23f27c148a43 for the line numbers, I don't expect these to happen under normal circumstances)
- put a `g_message("blocked device, not handling event")` on line 31 of `AbstractInputHandler.cpp`. 

Then report back what new messages you see, while using the build with these debug messages.
You could later also add some debug messages in `SetsquareInputHandler.cpp` (current lines 29, 46, 117, 194).

I really don't understand the problem and only debug messages can help to find the issue.

lxdb commented:
@rolandlo Sorry, in the last few days I didn't have much time. Maybe today I'll be able to do it.

lxdb commented:
https://user-images.githubusercontent.com/47871822/155877619-b044a8ef-f38c-437e-9d32-6f30d4624d12.mp4

@rolandlo Here you can see what happens. The messages come every time I use the pen.

PeterPistazius commented:
I love to see this tool. Thank you so much for your effort! Helps me out a lot since i do a lot of teaching online.

I was wondering if it is possible to add a little slider knob at the tip of the setsquare. So you can position the setsquare with one finger perfectly in place, and then change the angle with the slider knob. 

Another idea was to open a text input by pressing the angle on the setsquare, and enter an angle manually.

What do you think about the ideas?
Idea 1 could look something like that:
![image](https://user-images.githubusercontent.com/124735540/217364056-a9563e5d-a9cd-4eb5-88c6-5e65ab8b48dc.png)

If this is not the right place for that i apolize

rolandlo commented:
@PeterPistazius
Thanks for your feedback and for the suggestions. I'm glad you love to see the tool. There seems to be consensus on the need to have better fine-tuned control over the rotation angle (c.f. https://github.com/xournalpp/xournalpp/discussions/4254#discussioncomment-4059865). Your suggestion (slider knob) looks promising, as it might not interfere too much with what is already there.
The right place to discuss this would be in the discussions, either in https://github.com/xournalpp/xournalpp/discussions/4254 or (maybe preferably) in a new discussion.
Some points that would have to be discussed as well:
- How would you expect the slider knob be manipulated (it would be easier to implement via stylus/mouse than via touch)?
- Is that a setsquare-only thing or would it generalize to other geometry tools (in particular to the compass)? 

By the way as of right now the way to fine-tune the rotation angle that works best for me is using the keyboard (Alt+r/Alt+R, which results in 0.2 degree increments). The keyboard is not available to everyone and is mostly meant to replace touch for those who don't have a touch device. So a solution using only touch and/or pointer device (mouse/stylus) is still needed.
